File: 684.RedundantConnection.py
#DFS Solution. T : O(n^2)

class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        Map = {e:[] for e in range(1, len(edges) + 1)}

        visit = set()
        def dfs(start, end):
            if end in Map[start]:
                return False
            if start in visit:
                return True
            visit.add(start)
            for c in Map[start]:
                if not dfs(c, end):
                    return False
            visit.remove(start)
            return True
        for s, e in edges:
            if dfs(s, e):
                Map[s].append(e)
                Map[e].append(s)
            else:
                return [s, e]

File: 684.RedundantConnection.go
func findRedundantConnection(edges [][]int) []int {
    n := len(edges)
    par := make( []int, n + 1)
    rank := make( []int, n + 1)

    
    find := func(node int) int {
        p := par[node]
        for p != par[p]{
            par[p] = par[par[p]]
            p = par[p]
        }
        return p
    }

    
    union := func(node1 int, node2 int) bool{
        p1, p2 := find(node1), find(node2)

        if p1 == p2{
            return false
        }

        if rank[p1] > rank[p2]{
            par[p2] = p1
            rank[p1] += rank[p2]
        }else{
            par[p1] = p2
            rank[p2] += rank[p1]
        }
        return true
    }

    for i := 0; i < n + 1; i++{
        par[i] = i
        rank[i] = 1
    }

    for _, edge := range edges{
        n1, n2 := edge[0],edge[1]
        if union(n1, n2) == false{
            return edge
        }
        
    }

    return []int{}
}
